src/libcsg/modules/io/xmltopologyreader.cc

```diff
@@ -24,152 +24,319 @@
 
-bool XMLTopologyReader::ReadTopology(string filename, Topology &top)
-{ 
-  _top = &top;
+bool XMLTopologyReader::ReadTopology(string filename, Topology &top)  {
+    _top = &top;
 
-  _parser.NextHandler(this, &XMLTopologyReader::ParseRoot);
-  _parser.Open(filename);
-    
-  return true;
+    Property options;
+    load_property_from_xml(options, filename);
+    ParseRoot(options.get("topology"));
+
+    _top->RebuildExclusions();
+    return true;
 }
 
-void XMLTopologyReader::ReadTopolFile(string file)
-{
+void XMLTopologyReader::ReadTopolFile(string file) {
     TopologyReader *reader;
     reader = TopReaderFactory().Create(file);
     if(!reader)
         throw runtime_error(file + ": unknown topology format");
     
     reader->ReadTopology(file, *_top);
-    
+
     delete reader;
+    // Clean XML molecules and beads.
 }
 
-void XMLTopologyReader::ParseRoot(const string &el, map<string, string> &attr)
-{
-    if(el == "topology") {
-        if(attr["base"] != "")
-            ReadTopolFile(attr["base"]);
-        _parser.NextHandler(this, &XMLTopologyReader::ParseTopology);
+void XMLTopologyReader::ParseRoot(Property &property)  {
+    if (property.hasAttribute("base")) {
+        ReadTopolFile(property.getAttribute<string>("base"));
     }
-    else {
-        throw std::runtime_error("wrong root node in xml topology file");
+
+    // Iterate over keys at first level.
+    for (Property::iterator it = property.begin(); it != property.end(); ++it) {
+        if (it->name() == "h5md_particle_group") {
+            _top->setParticleGroup(it->getAttribute<string>("name"));
+        } else if (it->name() == "molecules") {
+            _mol_index = 1;
+            _bead_index = 0;
+            ParseMolecules(*it);
+        } else if (it->name() == "bonded") {
+            ParseBonded(*it);
+        } else if (it->name() == "box") {
+            ParseBox(*it);
+        } else if (it->name() == "beadtypes") {
+            ParseBeadTypes(*it);
+        } else {
+            throw runtime_error("unknown tag: topology."+ it->name());
+        }
     }
 }
 
-void XMLTopologyReader::ParseTopology(const string &el, map<string, string> &attr)
-{
-    if(el == "molecules") {
-         _parser.NextHandler(this, &XMLTopologyReader::ParseMolecules);
-    }
-    else if(el == "box") {
-        matrix m;
-        m.ZeroMatrix();
-        string xx = attr["xx"];
-        string yy = attr["yy"];
-        string zz = attr["zz"];
-        if (xx == "" || yy == "" || zz == "")
-            throw runtime_error("invalid box tag - missing xx, yy or zz");
-        try {
-           m[0][0] = boost::lexical_cast<double>(xx);
-        } catch(boost::bad_lexical_cast &) {
-           throw std::runtime_error("Cannot convert xx='"+ xx + "' to double");
+void XMLTopologyReader::ParseBox(Property &p) {
+    matrix m;
+    m[0][0] = p.getAttribute<double>("xx");
+    m[1][1] = p.getAttribute<double>("yy");
+    m[2][2] = p.getAttribute<double>("zz");
```

Comment:
Are the error messages still clear if one of them is missing and/or is not a double?